### PR TITLE
scatter_add_decomposition

### PR DESCRIPTION
**Pull request by apbose**
The issue in this PR is for cases where there is index collision. Example of such cases-
`scatter_add(input_tensor = torch.zeros(3,5), dim=1, index_tensor = torch.tensor([[0, 1, 2, 0]]).cuda(), src_tensor = torch.tensor([[1, 2, 3, 1]], dtype=torch.int32).cuda())`
Looks like straighforward decomposition by stacking the src_tensor would not work, since it would again lead to collision for some cases.
The better way to implement is to do converter implementation and allot the values using advanced indexing.
Working on this now.

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 1]
```diff
@@ -174,6 +174,19 @@ def empty_permuted_decomposition(*args, **kwargs) -> torch.Tensor:
...
+@register_torch_trt_decomposition(
+    torch.ops.aten.scatter_add.default, registry=TORCH_TRT_DECOMPOSITIONS
+)
+def scatter_add_decomposition(
+    input_tensor: torch.Tensor,
+    src_tensor: torch.Tensor,
+    dim: int,
+    index: torch.Tensor,
+) -> torch.Tensor:
+    input_tensor_to_add = torch.scatter(torch.empty_like(input_tensor), dim, index, src_tensor)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:186]: Could `torch.empty_like(input_tensor)` instead be just `input_tensor`, or is it required to be empty?

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 2]
```diff
@@ -174,6 +174,19 @@ def empty_permuted_decomposition(*args, **kwargs) -> torch.Tensor:
...
+    scatter_add_tensor = torch.add(input_tensor, input_tensor_to_add.cuda())
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:187]: Without `.cuda()` does the decomposition fail?

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 1 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py]: Yes it would be required. The logic is like we would apply scatter with the indices to an empty tensor and then add it to the input tensor.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 2 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py]: Yes in this case it complains that the two tensors are on different devices.

**Comment by apbose** [comment]
@gs-olive though the test cases pass, I don't think that the decomposition is taking place.
It fails in the backend at `aot_export_joint_simple`. Shows me something like

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 3]
```diff
@@ -243,6 +243,44 @@ def empty_strided_decomposition(*args, **kwargs) -> torch.Tensor:
...
 
+@register_torch_trt_decomposition(
+    torch.ops.aten.scatter_add.default, registry=TORCH_TRT_DECOMPOSITIONS
+)
+def scatter_add_decomposition(
+    input_tensor: torch.Tensor,
+    src_tensor: torch.Tensor,
+    dim: int,
+    index: torch.Tensor,
+) -> torch.Tensor:
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:254]: It seems the order of args doesn't match with tests, which made CI failed.

Comment (apbose) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py]: Thanks for pointing this out! Corrected.

**Review by zewenli98 — APPROVED** [review]
LGTM
